octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,348 @@
+    CHANGE_LOOP_WAIT = 5
+
+    def __init__(self, id, project=None, credentials_file=None,
+                 *args, **kwargs):
+
+        if credentials_file:
+            self.gcloud_client = dns.Client.from_service_account_json(
+                credentials_file, project=project)
+        else:
+            self.gcloud_client = dns.Client(project=project)
+
+        # Logger
+        self.log = getLogger('GoogleCloudProvider[{}]'.format(id))
+        self.id = id
+
+        self._gcloud_zones = {}
+
+        super(GoogleCloudProvider, self).__init__(id, *args, **kwargs)
+
+    def _apply(self, plan):
+        """Required function of manager.py to actually apply a record change.
+
+            :param plan: Contains the zones and changes to be made
+            :type  plan: octodns.provider.base.Plan
+
+            :type return: void
+        """
+        desired = plan.desired
+        changes = plan.changes
+
+        self.log.debug('_apply: zone=%s, len(changes)=%d', desired.name,
+                       len(changes))
+
+        # Get gcloud zone, or create one if none existed before.
+        if desired.name not in self.gcloud_zones:
+            gcloud_zone = self._create_gcloud_zone(desired.name)
+        else:
+            gcloud_zone = self.gcloud_zones.get(desired.name)
+
+        gcloud_changes = gcloud_zone.changes()
+
+        for change in changes:
+            class_name = change.__class__.__name__
+            _rrset_func = getattr(
+                self, '_rrset_for_{}'.format(change.record._type))
+
+            if class_name == 'Create':
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Delete':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Update':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.existing))
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.new))
+            else:
+                raise RuntimeError('Change type "{}" for change "{!s}" '
+                                   'is none of "Create", "Delete" or "Update'
+                                   .format(class_name, change))
+
+        gcloud_changes.create()
+
+        for i in range(120):
+            gcloud_changes.reload()
+            self.log.debug("Waiting for changes to complete")
+            # https://cloud.google.com/dns/api/v1/changes#resource
+            # status can be one of either "pending" or "done"
+            if gcloud_changes.status != 'pending':
+                break
+            self.log.debug("Waiting for changes to complete")
+            time.sleep(self.CHANGE_LOOP_WAIT)
+
+        if gcloud_changes.status != 'done':
+            raise RuntimeError("Timeout reached after {} seconds".format(
+                i * self.CHANGE_LOOP_WAIT))
```

Comment:
Will this always be a timeout? Is it possible that `status == 'error'`?

Reply:
status is one of either "done" or "pending". Errors raises exceptions.